### PR TITLE
MIDI Channel A/B

### PR DESCRIPTION
**Pull request by milkmiruku**
Added a new Carla internal plugin that sends each MIDI channel to either the A or B output (#621)

### CODE REVIEW

File: source/native-plugins/midi-channel-ab.c [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+{
+    MidiGainHandle* const handle = (MidiGainHandle*)malloc(sizeof(MidiGainHandle));
+
+    if (handle == NULL)
+        return NULL;
+
+    handle->host = host;
+
+    for (int i=MAX_MIDI_CHANNELS; --i>=0;)
+        handle->channels[i] = 0.0f;
```
Comment (falkTX) [inline source/native-plugins/midi-channel-ab.c]: This is not right, the channels are booleans, not floats.
And to zero memory we can use `memset(ptr, 0, full-size)` directly, instead of going through a list.

File: source/native-plugins/midi-channel-ab.c [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+{
+    return MAX_MIDI_CHANNELS;
+
+    // unused
+    (void)handle;
+}
+
+static const NativeParameter* midichanab_get_parameter_info(NativePluginHandle handle, uint32_t index)
+{
+    if (index > MAX_MIDI_CHANNELS)
```
Comment (falkTX) [inline source/native-plugins/midi-channel-ab.c]: This needs to be >=

File: source/native-plugins/midi-channel-ab.c [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+
+    return &param;
+
+    // unused
+    (void)handle;
+}
+
+static float midichanab_get_parameter_value(NativePluginHandle handle, uint32_t index)
+{
+    if (index > MAX_MIDI_CHANNELS)
```
Comment (falkTX) [inline source/native-plugins/midi-channel-ab.c]: Same, >=

File: source/native-plugins/midi-channel-ab.c [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+        return 0.0f;
+
+    return handlePtr->channels[index] ? 1.0f : 0.0f;
+}
+
+static void midichanab_set_parameter_value(NativePluginHandle handle, uint32_t index, float value)
+{
+    if (index > MAX_MIDI_CHANNELS)
```
Comment (falkTX) [inline source/native-plugins/midi-channel-ab.c]: Same, >=

File: source/native-plugins/midi-channel-ab.c [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+        if (MIDI_IS_CHANNEL_MESSAGE(status))
+        {
+            const uint8_t channel = (uint8_t)MIDI_GET_CHANNEL_FROM_DATA(midiEvent->data);
+
+            if (channel >= MAX_MIDI_CHANNELS)
+                continue;
+
+            if (channels[channel])
+            {
+                tmpEvent.port    = midiEvent->port+1;
```
Comment (falkTX) [inline source/native-plugins/midi-channel-ab.c]: this part is needlessly extensive.
please do a `memcpy(&tmpEvent, midiEvent, sizeof(NativeMidiEvent));` and then only modify the bits needed.

File: source/native-plugins/midi-channel-ab.c [hunk 6]
```diff
@@ -92,18 +91,18 @@ static const NativeParameter* midichanab_get_parameter_info(NativePluginHandle h
...
 }
 
 static void midichanab_set_parameter_value(NativePluginHandle handle, uint32_t index, float value)
 {
-    if (index > MAX_MIDI_CHANNELS)
+    if (index >= MAX_MIDI_CHANNELS)
         return;
 
-    handlePtr->channels[index] = (value >= 0.5f);
+    handlePtr->channels[index] = (value >= 1);
```
Comment (falkTX) [inline source/native-plugins/midi-channel-ab.c]: why did you change all floats to integers here?
the actual type is a float, so the *.0f makes sense, it is easy to understand when reading the code that the actual type is a float and not an integer.
The specific change that this comment replies to, please change it back to >= 0.5f.

Comment (milkmiruku) [reply source/native-plugins/midi-channel-ab.c]: Ah, thank you. I'm not familiar with how to read C++.

File: source/native-plugins/_data.all.cpp [hunk 7]
```diff
@@ -83,6 +83,22 @@ static const NativePluginDescriptor sNativePluginDescriptors[] = {
...
     DESCFUNCS
 },
+{
+    /* category  */ NATIVE_PLUGIN_CATEGORY_UTILITY,
+    /* hints     */ NATIVE_PLUGIN_IS_RTSAFE,
+    /* supports  */ NATIVE_PLUGIN_SUPPORTS_EVERYTHING,
+    /* audioIns  */ 0,
+    /* audioOuts */ 0,
+    /* midiIns   */ 1,
+    /* midiOuts  */ 1,
```
Comment (falkTX) [inline source/native-plugins/_data.all.cpp]: Typo here, needs to be 2, not 1

File: source/native-plugins/_data.base.cpp [hunk 8]
```diff
@@ -67,6 +67,22 @@ static const NativePluginDescriptor sNativePluginDescriptors[] = {
...
     DESCFUNCS
 },
+{
+    /* category  */ NATIVE_PLUGIN_CATEGORY_UTILITY,
+    /* hints     */ NATIVE_PLUGIN_IS_RTSAFE,
+    /* supports  */ NATIVE_PLUGIN_SUPPORTS_EVERYTHING,
+    /* audioIns  */ 0,
+    /* audioOuts */ 0,
+    /* midiIns   */ 1,
+    /* midiOuts  */ 1,
```
Comment (falkTX) [inline source/native-plugins/_data.base.cpp]: Same, needs to be 2

File: source/native-plugins/_data.all.cpp [hunk 9]
```diff
@@ -83,6 +83,22 @@ static const NativePluginDescriptor sNativePluginDescriptors[] = {
...
+    /* supports  */ NATIVE_PLUGIN_SUPPORTS_EVERYTHING,
+    /* audioIns  */ 0,
+    /* audioOuts */ 0,
+    /* midiIns   */ 1,
+    /* midiOuts  */ 2,
+    /* paramIns  */ 0,
+    /* paramOuts */ 0,
+    /* name      */ "MIDI Channel A/B",
+    /* label     */ "midichanab",
+    /* maker     */ "falkTX",
```
Comment (falkTX) [inline source/native-plugins/_data.all.cpp]: You have maker here as me, but the actual plugin code has it as "Milk Brewster".
Which one is it?
If you want to set yourself as the author, that is completely fine, but please change in all places on the PR and Update the header comments too regarding it.
No need for "Copyright (C) 2012-2015 Filipe Coelho", just use your usual one-liner for these things.
I assume you're okay with GPL v2+ ?

Comment (milkmiruku) [reply source/native-plugins/_data.all.cpp]: Thanks for catching that. Unless I'm mistaken, it's corrected below.

I've left your copyright header in midi-channel-ab.c also, given it's based on your code. Should I for sure be adding a line to files other than midi-channel-ab.c? I don't mind either way.

I'm fine with GPL v2+

**Comment by falkTX** [comment]
Looks good, will merge now, thanks!
